wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -90,80 +129,124 @@ impl LastVotedForkSlotsAggregate {
         &mut self,
         new_slots: RestartLastVotedForkSlots,
     ) -> Option<LastVotedForkSlotsRecord> {
-        let total_stake = self.epoch_stakes.total_stake();
-        let threshold_stake = (total_stake as f64 * self.repair_threshold) as u64;
         let from = &new_slots.from;
         if from == &self.my_pubkey {
             return None;
         }
-        let sender_stake = Self::validator_stake(&self.epoch_stakes, from);
-        if sender_stake == 0 {
-            warn!(
-                "Gossip should not accept zero-stake RestartLastVotedFork from {:?}",
-                from
-            );
+        let root_slot = self.root_bank.slot();
+        let new_slots_vec = new_slots.to_slots(root_slot);
+        if new_slots_vec.is_empty() {
             return None;
         }
-        self.active_peers.insert(*from);
-        let new_slots_vec = new_slots.to_slots(self.root_slot);
+        let last_vote_epoch = self
+            .root_bank
+            .get_epoch_and_slot_index(*new_slots_vec.last().unwrap())
+            .0;
+        if self
+            .node_to_last_vote_epoch_map
+            .insert(*from, last_vote_epoch)
+            != Some(last_vote_epoch)
+        {
+            self.update_epoch_info();
+        }
         let record = LastVotedForkSlotsRecord {
             last_voted_fork_slots: new_slots_vec.clone(),
             last_vote_bankhash: new_slots.last_voted_hash.to_string(),
             shred_version: new_slots.shred_version as u32,
             wallclock: new_slots.wallclock,
         };
+        if self.update_and_check_if_message_already_saved(new_slots, new_slots_vec) {
+            return None;
+        }
```

Comment:
This can be moved before we construct the `record` object above

Reply:
It is written like this so we can avoid a new_slots.clone(). But maybe a clone isn't that bad, changed.